gems/codetracer-ruby-recorder/ext/native_tracer/src/lib.rs

```diff
@@ -281,21 +269,46 @@ unsafe extern "C" fn disable_tracing(self_val: VALUE) -> VALUE {
     Qnil.into()
 }
 
-fn flush_to_dir(
-    tracer: &Tracer,
+fn begin_trace(
     dir: &Path,
     format: runtime_tracing::TraceEventsFileFormat,
-) -> Result<(), Box<dyn std::error::Error>> {
+) -> Result<Box<dyn TraceWriter>, Box<dyn std::error::Error>> {
+    let mut tracer = create_trace_writer("ruby", &vec![], format);
     std::fs::create_dir_all(dir)?;
     let events = match format {
         runtime_tracing::TraceEventsFileFormat::Json => dir.join("trace.json"),
-        runtime_tracing::TraceEventsFileFormat::Binary => dir.join("trace.bin"),
+        runtime_tracing::TraceEventsFileFormat::BinaryV0
+        | runtime_tracing::TraceEventsFileFormat::Binary => dir.join("trace.bin"),
     };
     let metadata = dir.join("trace_metadata.json");
     let paths = dir.join("trace_paths.json");
-    tracer.store_trace_events(&events, format)?;
-    tracer.store_trace_metadata(&metadata)?;
-    tracer.store_trace_paths(&paths)?;
+
+    TraceWriter::begin_writing_trace_events(&mut *tracer, &events)?;
+    TraceWriter::begin_writing_trace_metadata(&mut *tracer, &metadata)?;
+    TraceWriter::begin_writing_trace_paths(&mut *tracer, &paths)?;
+
+    Ok(tracer)
+}
+
+fn flush_to_dir(
+    tracer: &mut dyn TraceWriter,
+    //    dir: &Path,
+    //    format: runtime_tracing::TraceEventsFileFormat,
+) -> Result<(), Box<dyn std::error::Error>> {
+    //std::fs::create_dir_all(dir)?;
+    //let events = match format {
+    //    runtime_tracing::TraceEventsFileFormat::Json => dir.join("trace.json"),
+    //    runtime_tracing::TraceEventsFileFormat::BinaryV0 |
+    //    runtime_tracing::TraceEventsFileFormat::Binary => dir.join("trace.bin"),
```

Comment:
maybe the old commented out code can be directly removed now, if the new one works